### PR TITLE
Feature/appdev 1123 build up on jenkins ext

### PR DESCRIPTION
**Pull request by vishal-revdiwala**
Branch is Enabled with Jenkins file for build automation (CI).
Removed non working old YonaUITests from target
Removed non working old unit tests from target
Removed existing old pods from repo. 
Jenkins Pipeline:
**Build Stage**
    - Clean existing workspace
    - Checkout scm
    - Install Project Dependencies ( Cocoapods ) 
    - Versioning the Build with avgtool. This will increment the build number by one.
    - Create Archive for Debug(Development) and Release(App-Store)
    - Export Archive to make both .ipa's
    - Staging the modified files after avgtool command.
    - Commit updated versioning files to repo branch 
    - Tags the local git in workspace.
    - Pushes the tag to Git repo.
   - Archiving Artifacts.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+          sh 'xcodebuild -exportArchive -archivePath ./BuildOutput/Yona-Release-$BUILD_NUMBER.xcarchive -exportPath ./BuildOutput/Yona-Release-$BUILD_NUMBER.ipa -exportOptionsPlist ./BuildOutput/ExportOptionsRelease.plist'
+          sh 'git checkout $BRANCH_NAME'
+          sh 'git add -u'
+          sh 'git commit -m "Updated versionCode for build $BUILD_NUMBER [ci skip]"'
+          sh 'git push https://${GIT_USR}:${GIT_PSW}@github.com/yonadev/yona-app-ios.git'
+          sh 'git tag -a $BRANCH_NAME-build-$BUILD_NUMBER -m "Jenkins"'
+          sh 'git push https://${GIT_USR}:${GIT_PSW}@github.com/yonadev/yona-app-ios.git --tags'
+        }
+        
+        archiveArtifacts(artifacts: 'Yona/BuildOutput/**/*.ipa', allowEmptyArchive: true)
```
Comment (aristotelos) [inline Jenkinsfile]: Are there cases in which the archive can be empty?

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        checkout scm
+        dir(path: 'Yona') {
+          sh '/usr/local/bin/pod install'
+          sh 'xcrun agvtool new-marketing-version 1.1'
+          sh 'xcrun agvtool next-version -all'
+          sh 'xcodebuild -allowProvisioningUpdates -workspace Yona.xcworkspace -configuration Debug -scheme Yona archive -archivePath ./BuildOutput/Yona-Debug-$BUILD_NUMBER.xcarchive'
+          sh 'xcodebuild -exportArchive -archivePath ./BuildOutput/Yona-Debug-$BUILD_NUMBER.xcarchive -exportPath ./BuildOutput/Yona-Debug-$BUILD_NUMBER.ipa -exportOptionsPlist ./BuildOutput/ExportOptionsDebug.plist'
+          sh 'xcodebuild -allowProvisioningUpdates -workspace Yona.xcworkspace -configuration Release -scheme Yona archive -archivePath ./BuildOutput/Yona-Release-$BUILD_NUMBER.xcarchive'
+          sh 'xcodebuild -exportArchive -archivePath ./BuildOutput/Yona-Release-$BUILD_NUMBER.xcarchive -exportPath ./BuildOutput/Yona-Release-$BUILD_NUMBER.ipa -exportOptionsPlist ./BuildOutput/ExportOptionsRelease.plist'
+          sh 'git checkout $BRANCH_NAME'
```
Comment (aristotelos) [inline Jenkinsfile]: Why is this necessary? The `checkout scm` command above already checks out the repository

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      }
+      environment {
+        GIT = credentials('65325e52-5ec0-46a7-a937-f81f545f3c1b')
+      }
+      steps {
+        deleteDir()
+        checkout scm
+        dir(path: 'Yona') {
+          sh '/usr/local/bin/pod install'
+          sh 'xcrun agvtool new-marketing-version 1.1'
```
Comment (aristotelos) [inline Jenkinsfile]: What if we leave this command out? The current version is already 1.1 isn't it, so we don't have to update it with avgtool?

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+          sh 'xcrun agvtool next-version -all'
+          sh 'xcodebuild -allowProvisioningUpdates -workspace Yona.xcworkspace -configuration Debug -scheme Yona archive -archivePath ./BuildOutput/Yona-Debug-$BUILD_NUMBER.xcarchive'
+          sh 'xcodebuild -exportArchive -archivePath ./BuildOutput/Yona-Debug-$BUILD_NUMBER.xcarchive -exportPath ./BuildOutput/Yona-Debug-$BUILD_NUMBER.ipa -exportOptionsPlist ./BuildOutput/ExportOptionsDebug.plist'
+          sh 'xcodebuild -allowProvisioningUpdates -workspace Yona.xcworkspace -configuration Release -scheme Yona archive -archivePath ./BuildOutput/Yona-Release-$BUILD_NUMBER.xcarchive'
+          sh 'xcodebuild -exportArchive -archivePath ./BuildOutput/Yona-Release-$BUILD_NUMBER.xcarchive -exportPath ./BuildOutput/Yona-Release-$BUILD_NUMBER.ipa -exportOptionsPlist ./BuildOutput/ExportOptionsRelease.plist'
+          sh 'git checkout $BRANCH_NAME'
+          sh 'git add -u'
```
Comment (aristotelos) [inline Jenkinsfile]: What files does this `git add` command actually add to git? Adding binary artifacts to git is not a good practice.

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+          sh 'git commit -m "Updated versionCode for build $BUILD_NUMBER [ci skip]"'
+          sh 'git push https://${GIT_USR}:${GIT_PSW}@github.com/yonadev/yona-app-ios.git'
```
Comment (aristotelos) [inline Jenkinsfile]: We can leave this first push out, because we can push the commit and the tag together two lines below.

File: Yona/BuildOutput/ExportOptionsDebug.plist [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (aristotelos) [inline Yona/BuildOutput/ExportOptionsDebug.plist]: Can't we move this file outside the `BuildOutput` directory? It is not build output but build input...

File: Yona/BuildOutput/ExportOptionsRelease.plist [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (aristotelos) [inline Yona/BuildOutput/ExportOptionsRelease.plist]: Same here: Can't we move this file outside the `BuildOutput` directory? It is not build output but build input...

File: Yona/Yona/Info.plist [hunk 8]
```diff
@@ -75,6 +75,7 @@
 	<key>UIRequiredDeviceCapabilities</key>
 	<array>
 		<string>armv7</string>
+		<string>gamekit</string>
```
Comment (aristotelos) [inline Yona/Yona/Info.plist]: Why is this capability required?

File: Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+          result = sh (script: "git log -1 | grep '.*\\[ci skip\\].*'", returnStatus: true) // Check if commit message contains skip ci label
+          result != 0 // Evaluate the result
+        }
+        
+      }
+      environment {
+        GIT = credentials('65325e52-5ec0-46a7-a937-f81f545f3c1b')
+      }
+      steps {
+        deleteDir()
```
Comment (aristotelos) [inline Jenkinsfile]: Are you sure this cleanup is necessary? I think every build starts clean automatically...

File: Jenkinsfile (hunk 3 above)
Comment (Bert-R) [reply Jenkinsfile]: I'd rather maintain the version number here, so we have a single place to look at, using the same approach for both Android and iOS

File: Jenkinsfile (hunk 2 above)
Comment (Bert-R) [reply Jenkinsfile]: Without this check-out, you can't check-in. I faced that for the Android build.

**Comment by aristotelos** [comment]
"Removed non working old unit tests from target" -> What unit tests were removed exactly?

File: Jenkinsfile (hunk 4 above)
Comment (Bert-R) [reply Jenkinsfile]: This adds the files that are updated through the ``avgtool`` invocations.

File: Jenkinsfile (hunk 5 above)
Comment (Bert-R) [reply Jenkinsfile]: Don't think so. The ``--tags`` option pushes the tags only.

File: Yona/Yona.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -134,7 +135,6 @@
 		9054CC841CE2270F00F575E2 /* RequestResult.swift in Sources */ = {isa = PBXBuildFile; fileRef = 9054CC831CE2270F00F575E2 /* RequestResult.swift */; };
 		9054CC861CE384A100F575E2 /* BuddyRequestManager.swift in Sources */ = {isa = PBXBuildFile; fileRef = 9054CC851CE384A100F575E2 /* BuddyRequestManager.swift */; };
 		9054CC881CE384AB00F575E2 /* Buddies.swift in Sources */ = {isa = PBXBuildFile; fileRef = 9054CC871CE384AB00F575E2 /* Buddies.swift */; };
-		9054CC8C1CE3882600F575E2 /* BuddyAPITests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 9054CC8B1CE3882600F575E2 /* BuddyAPITests.swift */; };
```
Comment (aristotelos) [inline Yona/Yona.xcodeproj/project.pbxproj]: Did we remove the reference but not the file here? Why?

File: Jenkinsfile (hunk 1 above)
Comment (Bert-R) [reply Jenkinsfile]: Shouldn't be. This flag should be ``false``. I corrected this for the Android build too.

**Comment by aristotelos** [comment]
"Removed non working old YonaUITests from target" -> This removes a lot of code indeed. Was all this code just example code from the internet that was not adapted for the Yona app? It has a strange copyright notice. So there was just a lot of example code that could be removed?

File: Jenkinsfile (hunk 2 above)
Comment (Sivateja-CMI) [reply Jenkinsfile]: configuring the git plugin with to "checkout to local matching branch" seems to work without extra check out step. Other wise the checkout scm is resulting in a detached head.
If we want to continue with detached head branch, we may need to pull the latest changes to it after checkout.

Comment (Bert-R) [reply Jenkinsfile]: Good action. Please update the Android build accordingly.

File: Jenkinsfile (hunk 1 above)
Comment (vishal-revdiwala) [reply Jenkinsfile]: corrected in iOS.

File: Yona/Yona.xcodeproj/project.pbxproj (hunk 10 above)
Comment (vishal-revdiwala) [reply Yona/Yona.xcodeproj/project.pbxproj]: I've removed reference of unit test cases from target as they are not working currently. We have planned to fix them eventually.

File: Yona/Yona/Info.plist (hunk 8 above)
Comment (Sivateja-CMI) [reply Yona/Yona/Info.plist]: We don't need it for our application. I will remove this.  We can add it later if we need it.

File: Jenkinsfile (hunk 2 above)
Comment (Sivateja-CMI) [reply Jenkinsfile]: Done. Updated Android pipeline for develop branch. Pipeline executed without any issues after the commit.
I have done that directly without PR, as it is only one line change in jenkins file. Other change is in plugin configuration.

Comment (Bert-R) [reply Jenkinsfile]: Thanks!

File: Yona/Yona.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -2050,6 +1684,37 @@
 			shellPath = /bin/sh;
 			shellScript = "\"$PODS_ROOT/R.swift/rswift\" generate \"$SRCROOT\"";
 		};
+		E53085D12105CBCD00488CE0 /* ShellScript */ = {
```
[the remark this answers is not in the document]
Comment (vishal-revdiwala) [reply Yona/Yona.xcodeproj/project.pbxproj]: I've removed shell script from project.

File: Yona/BuildOutput/ExportOptionsDebug.plist (hunk 6 above)
Comment (Sivateja-CMI) [reply Yona/BuildOutput/ExportOptionsDebug.plist]: we've moved this out of the build output folder. We have added a Export options folder in GIT to store the plist's for each configuration

File: Yona/BuildOutput/ExportOptionsRelease.plist (hunk 7 above)
Comment (Sivateja-CMI) [reply Yona/BuildOutput/ExportOptionsRelease.plist]: we've moved this out of the build output folder. We have added a Export options folder in GIT to store the plist's for each configuration

File: Jenkinsfile (hunk 9 above)
Comment (Sivateja-CMI) [reply Jenkinsfile]: We have removed it now.
I had same doubt while putting this deleteDir() statment. But, we observed the that workspace is not cleaned up on every build in Mac slave.
so we added it as we have the artifacts files created during the build which are the cause of checkout failure. Now, we have modified plugin configuration to checkout the source to matching local branch and added the contents of Build output to gitignore. After this there is no need of this command.
